content/blog/2024-06-17-gsoc-multiboot2.mdx

```diff
@@ -0,0 +1,58 @@
+---
+title: "Multiboot2 Support in Unikraft"
+description: |
+  Multiboot2, an enhanced version of Multiboot, solves some of the limitations of the original protocol, such as the lack of support for 64-bit UEFI systems. It also provides a more accurate and detailed communication during the bootstrapping process and offers a standardized tag system for easier configuration.
+publishedDate: 2024-06-17
+tags:
+- GSoC'24
+- Multiboot2
+- Booting
+authors:
+- Maria Pana
+---
+
+<img width="100px" src="https://summerofcode.withgoogle.com/assets/media/gsoc-generic-badge.svg" align="right" />
+
+## GSoC'24: Multiboot2 Support in Unikraft
+
+### Project overview
+
+When booting an x86 system, the CPU starts in 16-bit mode and the memory is limited to 1MB (in Real Mode).
+In order to find the first instruction to execute, the CPU jumps to a specific memory address known as a reset vector (e.g. physical address 0xFFFF0 for Intel 8086 processors).
```

Comment:
Will do. The "e.g." stems from finding 0xFFFFF0 for Intel 80286 and 0xFFFFFFF0 for Intel 80386. Is the actual address relevant to the reset vector or is it just the "16 bytes under" idea? Thanks!

Reply:
Huh :sweat_smile:
So, 80286 is the old 16-bit only CPU.
80386 is the first 32-bit CPU from the x86 family.
The answer is kind of combined. It can be either just `0xFFFF0` or `0xFFFFFFF0` as well. This has to do with what is known as BIOS shadowing.
When you press the Power On button, it is hardwired that the BIOS starts executing directly from flash that is at address `0xFFFF0` (Memory cycles in range 0xF0000 -> 0xFFFFF go to flash). This technique is called XIP (eXecute In Place), as RAM does not exist yet, the RAM controller has to be initialized, it's as if you would be executing code directly from storage. Modern chipsets however also decode memory cycles toward the last 2 MB of the frst 4GB (FFE0_0000 -> FFFF_FFFF) as memory cycles towards this flash memory (Memory aliasing). So basically *0xFFFF0 == *0xFFFFFFF0. As the BIOS initializes stuff, it enables 32-bit mode and may jump to execute from that top 4GB instead, but it's aliased so it executes the same code, it's just that `eip` instead of pointing to some low 16-bit address it points to some higher 32-bit address. At this stage, it enables RAM and then it programs some chipset registers known as PAM (Programmable Attribute Map) and here it has the option to allow memory cycles (either read, write or both, it can choose!) that go to 0xF0000 - 0xFFFFF to go to RAM instead (thus disabling aliasing). And this is how the BIOS ends up copying itself to RAM to execute faster, it will basically do a memcpy(0xF0000, 0xFFF00000, ...) and then jump again at that low 16-bit address which is now (faster) RAM instead of (slower) flash. 
As you can imagine, '286 being only 16-bit, does not go through this whole ordeal.
So, it should always be `0xffff0` regardless of generation. It's just that starting with 32-bit there is some aliasing going on so that BIOS is able to run faster at some point.

Reply:
Huh, out of curiosity this morning I decided to recheck, because I vaguely remember having seen something among those lines in UEFI as well. And Wikipedia indeed says `0xFFFFFFF0` and some Stackoverflow answers [agree](<https://electronics.stackexchange.com/questions/509025/reason-for-x86-reset-vector-being-at-0xfffffff0-instead-of-0xffffffff>) and some [disagree](<https://stackoverflow.com/questions/25470971/hooking-x86-reset-vector#:~:text=The%20reset%20vector%20for%20any,in%20software%20to%20prevent%20it.>) :thinking:. So then I decided to really confirm by actually checking the manual again to make sure I am not misremembering. And, indeed, the chapte 10.1.4 of the third volume of Intel SDM's gives the answer:
```
The first instruction that is fetched and executed following a hardware reset is located at physical address
FFFFFFF0H. This address is 16 bytes below the processor’s uppermost physical address. The EPROM containing the
software-initialization code must be located at this address.
The address FFFFFFF0H is beyond the 1-MByte addressable range of the processor while in real-address mode. The
processor is initialized to this starting address as follows. The CS register has two parts: the visible segment
selector part and the hidden base address part. In real-address mode, the base address is normally formed by
shifting the 16-bit segment selector value 4 bits to the left to produce a 20-bit base address. However, during a
hardware reset, the segment selector in the CS register is loaded with F000H and the base address is loaded with
FFFF0000H. The starting address is thus formed by adding the base address to the value in the EIP register (that
is, FFFF0000 + FFF0H = FFFFFFF0H).
The first time the CS register is loaded with a new value after a hardware reset, the processor will follow the normal
rule for address translation in real-address mode (that is, [CS base address = CS segment selector * 16]). To
ensure that the base address in the CS register remains unchanged until the EPROM based software-initialization
code is completed, the code must not contain a far jump or far call or allow an interrupt to occur (which would
cause the CS selector value to be changed).
```
So it does indeed start in Unreal Mode, as far as addressing for instructions fetching goes, but the data segments are still very much 16-bit only. So then I checked the UEFI code. The [reset vector indeed starts 16-byte under 4G](<https://github.com/tianocore/edk2/blob/master/UefiCpuPkg/ResetVector/Vtf0/Ia16/ResetVectorVtf0.asm#L51>) and then flat addressing is set up for the [data segments as well](<https://github.com/tianocore/edk2/blob/master/UefiCpuPkg/ResetVector/Vtf0/Ia16/Real16ToFlat32.asm#L24>). I also rechecked the other chipset register stuff I mentioned (PAM) and that is indeed still true. So I have to correct myself here :nerd_face: :point_up:. The BIOS still does the same aforementioned steps, except that it starts directly at that higher address and when it sets up 32-bit, it only does so for data but not for code. Never make assumptions late at night based on what you think you remember :sweat_smile: always RTFM.

So, I guess, you can also add `0xFFFFFFF0` so that we can show off :rofl:.

Reply:
Wow, thank you for the clearing up! I'll update the reset vector value to illustrate all possible values for the x86 family to avoid any confusion :smile: